### ISSUE TITLE
Meta data string

### ISSUE THREAD

**Issue by rkluwen** [Viincenttt/MollieApi#132]
When I pass a string in the MetaData field of a PaymentRequest, it gets sql escaped.

Example:

Passing { "order_id": 1 } gets: "{\r\n \"order_id\":  1\r\n}

It gets difficult like that to parse the same meta data upon receiving it in a webhook.

Greets,

Rene Kluwen

**Comment by Viincenttt** [comment]
Hi rkluwen,

I am not sure I understand what you mean. From what I can see in your example, newline characters are added to the response, but the result is still the same right? You should still be able to retrieve the order_id value from the return response using any Json parser library. 

Kind regards,
Vincent

**Comment by rkluwen** [comment]
Also backslashes are added are being added in front of the quotes. Sorry, my example didn't make that clear.
"{\r\n \"order_id\": 1\r\n}
Right now in my webhook, I worked around it by stripslashing it.

**Comment by Viincenttt** [comment]
Do you have the same problem if you use the SetMetaData and GetMetaData methods?

**Comment by rkluwen** [comment]
I'm trying that at this moment.
I dion't use the GetMetaData method, because I just use the meta data in a php webhook.
But I changed the payment request code to use SetMetaData and will see if it makes a difference.
We will have to wait until a next real payment gets validated because the test interface doesn't support sepa requests.

**Comment by rkluwen** [comment]
Sorry for the late reply. I just got in the results of my test. Using SetMetaData yields even worse results. The slashes are added two times! This is what the Mollie interface says:

An example: "\"{\\r\\n  \\\"order_id\\\": \\\"201900082\\\",\\r\\n  \\\"emailname\\\": \\\"xxxxx\\\\nuser@domain.nl\\\",\\r\\n  \\\"returnpath\\\": \\\"urlpath/201900082\\\"\\r\\n}\""

I'm pretty sure those slashes are not there, when I set the data. And \r\n are a normal linefeed and carriage return.

**Comment by fguemara** [comment]
I had the same problem, & trying to manage this way:
`var webhookParams = JsonConvert.DeserializeObject<WebhookParams>(id.Replace(@"\", "")); `

**Comment by rkluwen** [comment]
As a workaround, I removed the slashes in the webhook. One time, I will find time to do a PR.

**Comment by rkluwen** [comment]
It's a  quirk in the json library being in use: JsonConvert.SerializeObject(metadataObj, jsonSerializerSettings).
The jsonSerializerSettings defaults to StringEscapeHandling  = Default: "Only control characters (e.g. newline) are escaped.". But apparantly also the quotes, in case you pass a string.

@fguemara: Your solution is what I read elsewhere as well.

I think what happens is this: MetaData is a string object of the PaymentRequest class. In BaseMollieClient::SendHttpRequest() (line 33), this request is serialized once again, which makes all the slashes in the string, being escaped.

A solution would be to make MetaData of type object, in stead of a string.

**Comment by Viincenttt** [comment]
I am still not able to reproduce this issue. This is one of the tests in the `PaymentTests.cs` file that should catch this issue:
```
[Test]
        public async Task CanCreatePaymentWithJsonMetaData() {
            // If: We create a payment with meta data
            string json = "{\"order_id\":\"4.40\"}";
            PaymentRequest paymentRequest = new PaymentRequest() {
                Amount = new Amount(Currency.EUR, "100.00"),
                Description = "Description",
                RedirectUrl = this.DefaultRedirectUrl,
                Metadata = json
            };

            // When: We send the payment request to Mollie
            PaymentResponse result = await this._paymentClient.CreatePaymentAsync(paymentRequest);

            // Then: Make sure we get the same json result as metadata
            Assert.AreEqual(json, result.Metadata);
        }
```

But this test passes without issues.

**Comment by fguemara** [comment]
In my case, I used the SetMetadata() method after initializing the paymentRequest.
SetMetadata() accepts an object, so I passed an object without any serializer.
The \ is added when the metadata is received back via the webhook.
You should add another test with SetMetadata().

**Comment by kryptonite** [comment]
I looked into this a while back and saw the same problem. The problem is not evident in the tests because the tests use the same library. The problem arises when you want to parse the result in another programming language/framework. I believe the OP was using PHP.

You can recreate the problem by inspecting the raw json messages the library sends.

When using the SetMetadata method, the object you pass get serialized and put into the string MetaData of the model that gets sent. That model is then serialized again when sending it to mollie. The screenshot bellow shows the json body that gets sent.

![mollie2](https://user-images.githubusercontent.com/1089576/70308554-d118ed00-180b-11ea-999b-508cbe58853d.PNG)

If you receive the response and parse it using this library it gets parsed correctly and using GetMetaData gets the correct result. But when you try to get this in another language/framework you run into issues.

The documentation says:

> Provide any data you like, for example a string or a JSON object. We will save the data alongside the payment. Whenever you fetch the payment with our API, we’ll also include the metadata. You can use up to approximately 1kB.

So sending an object will work. When I change to type of PaymentRequest.MetaData to object and just set the value directly:

```var model = new MetaDataModel();
            model.OrderId = 1;
            model.Email = "test@test.nl";
            
            var paymentRequest = new PaymentRequest()
            {
                Amount = new Amount(Currency.EUR, "100.00"),
                Description = "Test payment of the example project",
                RedirectUrl = "http://google.com",
                WebhookUrl = ""
                Metadata = model
            };
```

The request that is sent look like this:

![mollie4](https://user-images.githubusercontent.com/1089576/70309195-36211280-180d-11ea-87b9-bde5a8d6e9c5.PNG)

Wich looks correct.

I'm not familiar enough with the codebase to say we can just change .MetaData on all the models to object. I think this might lead to other problems down the road like people adding objects that are to complex to serialize.

**Comment by Viincenttt** [comment]
Thanks for the investigation, this certainly helps. Changing it to an object should not be a problem. However, we need to make sure this doesn't break json metadata that was previously stored in the incorrect format. I'll have to do some testing with this...

**Comment by Viincenttt** [comment]
This issue is now fixed in version 2.1.0.0. Take a look at the release notes for a full list of fixes and changes:
https://github.com/Viincenttt/MollieApi/releases/tag/2.1.0.0
